src/xrpld/app/misc/detail/LendingHelpers.cpp

```diff
@@ -0,0 +1,2019 @@
+        computeInterestAndFeeParts(latePaymentInterest, managementFeeRate);
+#endif
+    auto const [roundedLateInterest, roundedLateManagementFee] = [&]() {
+        auto const interest =
+            roundToAsset(asset, latePaymentInterest, loanScale);
+        return computeInterestAndFeeParts(
+            asset, interest, managementFeeRate, loanScale);
+    }();
+
+    XRPL_ASSERT(
+        roundedLateInterest >= 0,
+        "ripple::detail::computeLatePayment : valid late interest");
+    XRPL_ASSERT_PARTS(
+        periodic.specialCase != PaymentSpecialCase::extra,
+        "ripple::detail::computeLatePayment",
+        "no extra parts to this payment");
+    // Copy the periodic payment values, and add on the late interest.
+    // This preserves all the other fields without having to enumerate them.
+    PaymentComponentsPlus const late = [&]() {
+        auto inner = periodic;
+#if LOANCOMPLETE
+        inner.rawInterest += rawLateInterest;
+#endif
+
+        return PaymentComponentsPlus{
+            inner,
+            // A late payment pays both the normal fee, and the extra fees
+            periodic.untrackedManagementFee + latePaymentFee +
+                roundedLateManagementFee,
+            // A late payment increases the value of the loan by the difference
+            // between periodic and late payment interest
+            periodic.untrackedInterest + roundedLateInterest};
+    }();
+
+    XRPL_ASSERT_PARTS(
+        isRounded(asset, late.totalDue, loanScale),
+        "ripple::detail::computeLatePayment",
+        "total due is rounded");
+
+    if (amount < late.totalDue)
+    {
+        JLOG(j.warn()) << "Late loan payment amount is insufficient. Due: "
+                       << late.totalDue << ", paid: " << amount;
+        return Unexpected(tecINSUFFICIENT_PAYMENT);
+    }
+
+    return late;
+}
+
+/* Handle possible full payments.
+ *
+ * If this function processed a full payment, the return value will be
+ * a PaymentComponentsPlus object. Otherwise, it'll be an Unexpected with the
+ * error code the caller is expected to return. It should NEVER return
+ * tesSUCCESS
+ */
+
+Expected<PaymentComponentsPlus, TER>
+computeFullPayment(
+    Asset const& asset,
+    ApplyView& view,
+    Number const& principalOutstanding,
+    Number const& managementFeeOutstanding,
+    Number const& periodicPayment,
+    std::uint32_t paymentRemaining,
+    std::uint32_t prevPaymentDate,
+    std::uint32_t const startDate,
+    std::uint32_t const paymentInterval,
+    TenthBips32 const closeInterestRate,
+    std::int32_t loanScale,
+    Number const& totalInterestOutstanding,
+    Number const& periodicRate,
+    Number const& closePaymentFee,
+    STAmount const& amount,
+    TenthBips16 managementFeeRate,
+    beast::Journal j)
+{
+    if (paymentRemaining <= 1)
+        // If this is the last payment, it has to be a regular payment
+        return Unexpected(tecKILLED);
```

Comment:
A small improvement to help folks resolve issues.


```suggestion
    if (paymentRemaining <= 1)
    {
        // If this is the last payment, it has to be a regular payment
        JLOG(j.warn()) << "Last payment cannot be a full payment.";
        return Unexpected(tecKILLED);
    }
```